### PR TITLE
Ruby on Rails capstone project - Budget app

### PR DESCRIPTION
**Pull request by MozamelJawad**
**Hi Reviewer 👋**

In this project, I worked on the following features: 

- Splash screen
  - A simple page with the name of your app (yes, you need to choose one), and links to the sign up and log in pages.

- Sign up and log in pages
  - The user should be able to register in the app with full name, email and password (all mandatory).
  - The user can log into the app using email and password.
  - If the user is not logged in, they can't access pages that require the user to be logged in (all the pages described below).

- Home page (categories page)
  - When the user logs in, they are presented with the categories page.
  - For each category, the user can see their name, icon and the total amount of all the transactions that belongs to that category.
  - When the user clicks (or taps) on a category item, the application navigates to the transactions page for that category.
  - There is a button "add a new category" at the bottom that brings the user to the page to create a new category.

- Transactions page
  - For a given category, the list of transactions is presented, ordered by the most recent.
  - At the top of the page the user could see the total amount for the category (sum of all of the amounts of the transactions in that category).
  - There is a button "add a new transaction" at the bottom that brings the user to the page to create a new transaction.
  - When the user clicks on the "Back" button (<), the user navigates to the home page.

- "Add a new category" page
  - The user fills out a form to create a new category, indicating their name and icon (both mandatory).
  - The user clicks (or taps) the "Save" button to create the new category, and is taken to the home page on success.
  - When the user clicks on the "Back" button (<), the user navigates to the home page.

- "Add a new transaction" page
  - The user fills out a form to create a new transaction with:
  - name (mandatory)
  - amount (mandatory)
  - categories (mandatory at least one)
  - The user click (or taps) the "Save" button to create the new transaction, and is taken to the transactions page for that category.
  - When the user clicks on the "Back" button (<), the user navigates to the transactions page for that category.

- Testing requirements
  - Create unit and integration tests for all the most important components of your RoR application.

- Technical requirements
  - You should use Postgres as your database.
  - You should use devise for authentication.
  - You should validate all user input to make sure that anyone with bad intentions cannot compromise your app.
  - You can use a view template engine of your choice (.erb, .slim, .haml).
  - The project should be deployed and accessible online.

    [Like of Project Description ](https://www.loom.com/share/c0c7a103c4ee4c14ae0c2cd765c7208d)

### CODE REVIEW

File: spec/features/user_spec.rb [hunk 1]
```diff
@@ -0,0 +1,28 @@
+require 'rails_helper'
+
+RSpec.describe 'User', type: :feature do
+  before :all do
+    @user = User.create(name: 'new', email: 'new@gmail.com', password: '123123')
+  end
+
+  describe 'Log in page' do
+    it 'should log in with credentials' do
```
Comment (MahmoudBakr23) [inline spec/features/user_spec.rb:9]: Required

Please test the other features of the app as well, such as creating categories and expenses.

File: test/controllers/expenses_controller_test.rb [hunk 2]
```diff
@@ -0,0 +1,7 @@
+require 'test_helper'
+
+class ExpensesControllerTest < ActionDispatch::IntegrationTest
+  # test "the truth" do
```
Comment (MahmoudBakr23) [inline test/controllers/expenses_controller_test.rb]: Required
- Please make sure to not leave any empty file or unused code for a cleaner code base.

File: app/controllers/expenses_controller.rb [hunk 3]
```diff
@@ -0,0 +1,27 @@
+class ExpensesController < ApplicationController
+  before_action :authenticate_user!
+
+  def new
+    @expense = Expense.new
+    @categories = Category.all
```
Comment (MahmoudBakr23) [inline app/controllers/expenses_controller.rb]: Required

- Please make sure to solve the N+1 query problem as per the project requirements, consider using the includes method in order to do so.

File: spec/features/user_spec.rb [hunk 4]
```diff
@@ -0,0 +1,28 @@
+require 'rails_helper'
+
+RSpec.describe 'User', type: :feature do
```
Comment (MahmoudBakr23) [inline spec/features/user_spec.rb:3]: Required

- Please test the other features of the app as well, such as creating categories and expenses.

**Review by MahmoudBakr23 — CHANGES_REQUESTED** [review]
Hello @MozamelJawad 😃, 

Great work and effort you put into this project so far.

## Highlights 🚀 

There are some excellent points that you have made perfectly. Keep up! 💪
- Well done on creating the controllers' restful methods
- Great job with authentication so far
- Nice work with the models' validations
- Good work with the database setup, migrations, and the routes
- Great design
- Descriptive PR and README
- Nice presentation
- Correct Linters

However, there is always room for improvement.

> I'd like to remind you that your long hours of work are appreciated. However, the code review process is for improving your code, not criticizing it. 😄

 ## Required Changes ♻️
_Please check the comments under the review._

## Optional suggestions
_Every comment with the [OPTIONAL] prefix is not crucial enough to stop the approval of this PR. However, I strongly recommend you take them into account as they can make your code better._

N/A

Cheers, and Happy coding!👏👨‍💻💻

Feel free to mention @MahmoudBakr23 to leave any questions or comments in the PR thread if something is not 100% clear.

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Comment by MozamelJawad** [comment]
Thanks,@MahmoudBakr23, for your feedback.
I did the required changes.

**Review by cynthiainga — APPROVED** [review]
Hi @MozamelJawad, 

Your project is complete!  There is nothing else to say other than... it's time to merge it :shipit: 
Congratulations! 🎉 



## Highlights

- [ ] All changes are made ✔️ 
- [ ] Good design ✔️ 
- [ ] No linters errors ✔️ 
- [ ] Use Postgres as the database for the application ✔️ 
- [ ] Use `devise` for authentication ✔️ 

## Optional suggestions

_Every comment with the [OPTIONAL] prefix won't stop the approval of this PR. However, I strongly recommend you take them into account as they can make your code better. Some of them were missed by the previous reviewer and addressing them will really improve your application._

Cheers, and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please remember to tag me in your question so I can receive the notification.

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think the code review was unfair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Pull request merged**